### ISSUE TITLE
d3d12 please

### ISSUE THREAD

**Issue by niguzi** [bo3b/3Dmigoto#354]
Currently 3dmigoto does not support UE5 games, it uses d3d12.dll. In the near future, games like Grand Theft Auto VI will be released. I implore you to continue updating 3dmigoto to support UE5 games.

**Comment by leotorrez** [comment]
updating 3dmigoto to directx12 with proper support would likely take several maintainers and a few years of investment. Don't hold your breath on it :3

**Comment by DarkStarSword** [comment]
Dave has been working on geo-12, not sure what state it's currently at.

**Comment by leotorrez** [comment]
> Dave has been working on geo-12, not sure what state it's currently at.

What.... is geo-12? is it a branch/fork of 3dm?

**Comment by DarkStarSword** [comment]
Yes. geo-11 is a fork of 3DMigoto that replaces our dependency on nvidia 3D Vision, geo-12 is a work in progress DX12 port of that.

**Comment by SinsOfSeven** [comment]
Isn't Geo-12 is a ReShade plugin, which is used to add Geometric based Stereo 3D shaders, I was under the impression it was a spiritual successor, and was not aiming for feature parity. As of today, it does seem to be that way.

**Comment by DarkStarSword** [comment]
No, unless something has changed recently that I'm not aware of geo-12 is still a work in progress port of geo-11, though I don't see much recent activity on the geo-12 branch. @davegl1234 would be the best person to answer the current status, if he is allowed to say.

**Comment by SinsOfSeven** [comment]
Ah, I am. Thank you for the clarification. I had trouble tracking the Geo-12 project, and was only able to find things related to Geo-11. As Flugan's Geo3D is working for dx12 (I believe) I had assumed that must be it. 

I can understand not wanting to opensource Geo12 too early. I believe there will be a lot of challenges as dx11 and dx12 are functionally quite different even though the APIs are very similar at a surface level. I wish Dave the best if he is still working on it, I can't find any place where he posts any activity about it.

In your opinion, based on dx11 having a lot more abstraction, kind of just letting 3dm sit on top, and dx12 having a lot more bare-metal instructions, what are some of the major differences a 3dm12 or Geo12 would need to be wary of? When we've discussed the possibility in our community, we usually discuss how it would probably be easier to start from scratch, to avoid any hang-ups on these differences. I even worry that resource hashing would be very difficult for a dx12 equivalent.

**Comment by DarkStarSword** [comment]
Regarding the open source nature of geo-11 & geo-12 - there is a commercial entity involved that expressed some concerns over their competitors being able to use these projects without the same level of time/money investment as them. Now, in the case of geo-11 that source code really needs to be released since it is derived from 3DMigoto and is therefore subject to the terms of the GPLv3 and has already had a binary release and is therefore required to make the source code available to anyone who received a copy of that binary, and since it was a public release that is everyone.

The last communication we had on this subject they had agreed that the source for geo-11 would be released GPLv3, however our main contact at that company who was driving the project no longer works there and I'm not entirely sure who we should be talking to at the moment. We (@bo3b and myself) do have access to the source code and if necessary can release it ourselves, but I'm also reluctant to burn any bridges if they aren't expecting us to do that. We should probably put some pressure on them to make this happen - I personally haven't released anything using geo-11 in part because that would put me in an awkward position as long as the repository is private, and other modders who have released anything with geo-11 are already in an awkward position since they don't have the source code they are technically required to provide.

Initially it was uncertain whether geo-12 would be based on 3DMigoto or a new project written from scratch, however since I have access to that branch I can see that it has indeed been derived from the 3DMigoto/geo-11 code base, and therefore is subject to the same GPLv3 license as 3DMigoto/geo-11. However, since there has not yet been a release of geo-12 in any form, they are not required to provide source code for it yet (my preference for all of this would be that development for both projects happens in a public github repository like 3DMigoto).

**Comment by ThreeDeeJay** [comment]
@DarkStarSword Any idea if [RenoDX](https://github.com/clshortfuse/renodx) (shader replacer) alongside 3D Vision's native D3D12 support could be a viable alternative (at least on driver <=425.31 and <=2080 Ti)? https://github.com/clshortfuse/renodx/issues/303

**Comment by DarkStarSword** [comment]
I'm away taking care of sick family at the moment so can't really spend time on this right now, but just noting that Deviare is alternatively available under a commercial license and I'm 98% certain that Acer did purchase that, so that would not be usable as leverage.

But really it doesn't matter what license they are using Deviare under since their Geo12 branch that I have access to contains the COPYING.txt and LICENSE.GPL.txt showing that they forked 3DMigoto *AFTER* the change to GPLv3 only:

<img width="2233" height="424" alt="Image" src="https://github.com/user-attachments/assets/1ea4f0fd-2749-4ef5-a2c8-c9a1b9c1169c" />

<img width="2413" height="1455" alt="Image" src="https://github.com/user-attachments/assets/7991d854-73af-46a0-a0bd-4bc4a82bcf9f" />

<img width="851" height="341" alt="Image" src="https://github.com/user-attachments/assets/40592199-c8db-4ca4-9a38-306fd0b17a07" />

So, yes, I believe they are using 3DMigoto/geo-11/geo-12 under the terms of the GPLv3 and as such they are legally required to release the source code to anyone who has received a binary copy (i.e. if you own a Spatial Labs display and have launched any game that uses geo-11 / geo-12 such as Street Fighter 6 you received a binary copy and are therefore entitled to a copy of the source code - though note that does not cover the fix for the game which jointly belongs to Acer, whoever they contracted to create the fix, and the game developer/publisher. Only the source code to geo-11/geo-12 is covered by the GPLv3).

Strictly speaking, under section 8 paragraph 1 their license to use 3DMigoto has already been automatically terminated when they first released geo-11 publicly (years ago) but failed to comply with the license by not providing source code and failing to place appropriate open source notices in their software. This is not something Bo3b or I have done - this happened automatically when they failed to comply with the license agreement.

The GPLv3 section 8 provides two paragraphs ("forgiveness clauses") by which they can reinstate their license (this was one of the more significant changes from version 2 of the license, for which even accidental violation was effectively a legal death sentence - version 3 allows mistakes and oversights to be corrected by simply coming back into compliance, while still allowing permanent termination of the license if the parties continually fails to do so).

I have not as yet (and I don't think Bo3b has either) notified them of their violation, which means that the 30 day deadline specified in section 8 paragraph 3 to reinstate their license is not yet in effect - that's not something I particularly want to invoke without first trying to work this out amicably, although legally I don't believe we can actually notify them of their violation WITHOUT invoking that clause - the moment Bo3b or I send them a message on this subject they are under a 30 day count down to comply or else they burn that opportunity to reinstate their license under paragraph 3 (they could still reinstate it under paragraph 2, but that would be a risky move for them because of the whole permanent termination thing that paragraph 2 allows).

I'm afraid I've probably dropped the ball on this one - I just don't really have the motivation to follow through on any of this. Maybe someone else (Bo3b?) needs to drive this forwards, I'll help in any way I can.

**Comment by ThreeDeeJay** [comment]
Thanks, I really appreciate some transparency on the subject. 👍 

On a side note, I just noticed traces of [RE Framework](https://github.com/praydog/REFramework) in "geo-12"'s `dinput8.dll`.
<img width="539" height="398" alt="Image" src="https://github.com/user-attachments/assets/48d1a4be-da83-4073-82c8-a2e13112fb0b" />
I just hope it doesn't mean that that it can only add 3D to [RE Engine games](https://www.pcgamingwiki.com/wiki/Engine:RE_Engine), which Street Fighter 6 runs on by no coincidence, I presume. Because at that point we'd probably be better of just using [VRto3D](https://github.com/oneup03/VRto3D) or [WibbleWobbleVR](https://github.com/oneup03/VRto3D/wiki/WibbleWobbleVR3.0), though personally I prefer having the least amount of wrappers and layers of complexity, especially if those degrade performance and/or quality.

**Comment by bo3b** [comment]
> By the way [@bo3b](https://github.com/bo3b) for those of us who are still on driver 425.31 and can get 3D in Direct3D 12 games, did you ever come across any driver profile flags that could help 3D Vision Automatic fix shaders like shadows/lighting, perhaps by stereoizing everything (even if it comes at a performance cost)? And any idea [if RenoDX could assist in fixing shaders](https://github.com/clshortfuse/renodx/issues/303)?

No idea actually.  As far as we know the DX12 support in the driver was never finished, although we know of at least a couple of games, including Tomb Raider that used it successfully, so it wasn't that far off.  It's pretty unlikely they'd have added more flags for DX12 support, but I don't think anyone has looked at it in depth because of how limited it was. It's probably worth experimenting with the old driver and some DX12 games, by copy/paste a known DX12 profile like we used to for DX9.

In my experience 452.06 is better than 425.31, as long as you have the Schwing hack enabled.

I don't know about RenoDX, but taking a quick look there it seems like it's doing the same job as 3Dmigoto for replacing shaders and hooking games. I doesn't look like it provides enough value add to switch to that. It does DX12 shader hacking, but that's not the hard part. The hard part is generating stereo.

<br>

> I still haven't been able to wrap my head around why some games/apps look near perfect without any fixes, while others don't. Someone mentioned geo-11 only stereoizes vertex shaders, and games using deferred effects like shadows may remain 2D without manual fixes.

This happens because of how much variance there is in how games function internally.  Some games use a lot of old school graphic tricks to improve performance, typically by making assumptions about a 2D output to a screen. That breaks when we turn it into 3D.  

One of the biggest of these old tricks is deferred rendering, where they do some calculation in the PixelShader that normally would be in the VertexShader. Shadows were historically done via deferred rendering, which is where our techniques for deprojecting the vertices in the PixelShader and changing the depth came in.  Deferred rendering is implicitly assuming a 2D output window/port, which is why it fails in true 3D.

Now strangely, in modern versions of Unity, I've found that shadows are working correctly out of the box, and they don't appear to be using deferred rendering anymore. Deferred rendering was always a sick hack for 2D generation, and it doesn't work in VR for example. It got a lot of traction in graphic guru circles though, because it saved some performance.

geo-11 does the same approach that 3D Vision did, which is to stereoize all vertices in VertexShaders. There was no good way to handle deferred rendering in PixelShaders though, and you saw that even on DX9 games. NVidia's goal there was to make a universal fix via 3D Vision, but in practice it didn't work out because games are just too freakish in how they use the DX APIs.  That's where we have come in and fix stuff by hand. It's why we get better results than just 3D Vision, but of course it doesn't scale.

If a game doesn't do a bunch of hokey API tricks like deferred rendering, and just leaves the 3D pipeline unmolested, then you get perfect results when we stereoize, because everything is truly 3D.  Where it goes wrong is them trying to do hacks and tricks that break the 3D.  In VR you can't do these hacks or tricks either, and I think it's part of why things have improved recently.

<br>

> So to what extent did NVIDIA fix shaders on their/driver's end. and was anything ever done for D3D12 games at all? For instance, there's this app called Cavern that runs on Unity and [shadows are broken in D3D12](https://github.com/user-attachments/assets/e8c0c8ef-d91f-4746-a1fb-41bddbc73fdc) but as soon as I [switch to D3D11 they're magically fixed](https://github.com/user-attachments/assets/5319dc0d-4d4c-44c3-96d8-5a32f0c3225e). 🤔

NVidia did some fixing of shaders on their end, which is where the driver flags came in. They worked out some basic techniques that would work on a batch of games.  Our best know detail on the flags is still on [wiki.bo3b.net](https://wiki.bo3b.net/index.php?title=Driver_Profile_Settings)

They never had a good way to fix Unity games, which is why DarkStarSwords original work there was such a giant advancement, and leading to Unity Universal Fix, and UE4 Universal. I think the Unity Universal fix is now broken for any Unity from 2022 and newer, the graphic effects I see broken are different, and I don't see any of the regex match anymore. But I also see that shadows just magically work now with modern Unity versions like your Cavern example.  The failure in DX12 might be because of missing driver flags.

I don't think I ever saw any flags for DX12 games, but I also did not look closely. It would be worthwhile to inspect the driver profiles for any DX12 game that has StereoProfile set. There might be flags there no one has investigated.  A bit moot of course, because drivers are dead, but it's not impossible that a DX12 profile can bring 3D to a game that is otherwise not tagged with a stereo profile.

The driver flags are a set of heuristics that change how the 3D Vision driver would apply stereo to different parts of the graphics. For example, some games would use only square render targets/bitmaps for shadows, and they'd figure that out by debugging a setup, then set the flag to skip stereoizing those square targets.  The approach was OK, and kinda/sorta made sense at the time, as they were trying to distill some of the commonality in games down to some heuristic flags that could fix games and scale better than manually fixing shaders like we do.

It still has some value, because if the driver can fix some effects like this, it's less work for us to do. So we try to always start with best known setup.  geo-11 matches the heuristics of 3D Vision mostly, which is why older fixes tend to work so well with geo-11.  

<br> 

One of the weird things that took me forever to understand is that shadows in particular need to _not_ be stereoized.  Because they are projected into a scene, if they are stereoized like the geometry, then you get double shadows; not stereo shadows in the right spot.  It comes from the shadows being sort of bounced out of the scene like ray tracing. You don't see the actual shadow path, you see what it hits.  What it hits needs to be stereo, but the shadow itself is not.  In deferred shading, they just bounced this directly to 2D.

So some of the heuristics 3D Vision applies via driver flags is to handle that sort of idea.  Stereoize this, don't stereoize that.

<br>

> And in case you're still running 3D Vision yourself, have you looked at the possibility of adding 3D Vision Direct output to plugins like [UEVR](https://github.com/praydog/UEVR), [this plugin](https://github.com/Vital-Volkov/Stereoscopic-3D-system-for-Unreal-Engine-5) or other 3D plugins like [Geo3D](https://github.com/Flugan/Geo3D), [SuperDepth3D](https://github.com/BlueSkyDefender/Depth3D). Since Unreal Engine is open source, it would be great if we could expand its native [`-emulatestereo` launch mode](https://dev.epicgames.com/documentation/en-us/unreal-engine/unreal-engine-command-line-arguments-reference) to support 3DV Direct, especially for online games with anti-tamper/anti-cheat where plugins wouldn't be an option.

I made a prototype of UEVR to Leia output, but it was clunky and needed more work to get stereo sizing correct. Since then Oneup03 has made his excellent VRto3D, which combined with UEVR to make a sort of general purpose UE4+5 result. It's a little messy to set up of course, but the results are better because UEVR makes proper stereo to start. This also deprecates our UE4 Universal Fix, because in general UEVR works in more games, especially newer ones. And gives us a partial DX12 path. I could relatively easily add a Direct Mode output UEVR, but have only looked at it briefly.  The real trick there is getting the stereo enabled in a game that was not built with that turned on. That's the secret sauce of UEVR, so just enabling the stereo flags in UE is not sufficient, it also needs the stereo libraries and code that is stripped out by game devs.

There is also a working SuperDepthTo3DVision I made. I can't remember where it's at, maybe search mtbs3d forums. I released that a long time ago, but don't recall where. This allows you to use SuperDepth directly to DirectMode on 3D Vision hardware.  I seem to vaguely recall there was some licensing conflict, which is why I did not open-source it. 

I have not looked at these other plugins in any depth, but the real problem is lack of time to study all the variants out there. Shit is crazy complicated and it takes forever to understand what is necessary to get them to work. For Vital-Volkov for example- He's made 3D stereo output, but... UE4/5 does that natively if you set the flags internally. Which is what UEVR does to fake out games and make them stereo using the native UE render.  So I can't tell really what he adds? Does it work in games, or just new UE projects?  This is where everything is crazy complicated and there is only so much I can study, and I prefer to improve stuff that I already know will turn the dial for us.

One small example- I've just yesterday finally debugged a crash that would only happen with geo-11 on AMD. This took me 3 months of on and off study and debugging to figure out what happened, and I was only able to figure out by using Gemini AI suggestions. It knows about weird AMD requirements that I have no knowledge of.  The reason this was worth all that effort was because UE4 Universal Fix has been crashing on AMD, and it would be definitely nice to have it work. The answer was that geo-11 injects a new output parameter into a Geometry Shader, and that works fine with NVidia, but AMD is more strict and crashes. Now I know what happens, I can possibly fix other AMD crashes.

**Comment by oneup03** [comment]
> > I see. I always wondered how feasible it'd be to just piggy-back on UEVR (or even just the native `-emulatestereo` flag in case its SBS output is good enough) to display full res 3DV. But honestly, given how there's so many games/engines getting VR mods, it'd probably be a better idea to implement a straight-forward barebones VR driver (e.g. `openvr_api.dll` replacement or [OpenXR runtime like this](https://github.com/elliotttate/OpenXR-Simulator)) that just forwards the full res frames into 3DV Direct. VRto3D is a fantastic step in the right direction, but it's kinda hard to set up and maintain when I have an actual HMD I use for VR.
>

@ThreeDeeJay  Sounds like you still need to try the wibblewobble beta 😉 - I think this path to 3DVision hardware output without the buggy Nvidia driver is the way to go now. Full resolution per eye, mostly stays synchronized, and can run on any driver. 

Wibblewobble and VRto3D are very barebones. I don't think an OpenXR or native UEVR implementation would actually perform more than a few percent better. If you struggled to run newer games with VR mods, it's old CPU/GPU and not the VR driver hindering you. I don't think there's much point in creating an old 3DVision direct output for hardware that can't run the games well.

You're going to run into problems with OpenXR as well if you want to use a real HMD on the same PC as a 3D display. You will have to change the OpenXR runtime in your registry everytime you want to switch. With the current SteamVR drivers VRto3D and wibblewobble, you just enable/disable the add-ons in the SteamVR settings.

@bo3b I know you did some work on getting this Nvidia driverless 3DVision output working in Geo-11. I was able to take your emitter code updates and implement them in wibblewobble. I also added an EDID lookup on the NVTimingsEd database to use official timings. It's now able to consistently trigger 3DVision lightboost output automatically on any official 3DVision display, without crashes, and it's pretty 3D stable. If you'd like to copy some of that over to Geo-11 to replace (or supplement) the buggy DX9 mode, that would be awesome!

**Comment by ThreeDeeJay** [comment]
> mostly stays synchronized

By mostly, do you mean it occasionally desyncs/swaps eyes?
I remember that used to happen sometimes with my older 3DV hardware and it'd drive me nuts, so I'm wary of using non-native shuttering.

> If you struggled to run newer games with VR mods, it's old CPU/GPU and not the VR driver hindering you. I don't think there's much point in creating an old 3DVision direct output for hardware that can't run the games well.

Even with my aging hardware I can still manage to run VR games/mods with decent performance on mid-high settings with some tricks. The real problem I usually run into seems to be bottlenecks introduced by poor optimization/resource usage and/or multiple complexity layers where I'd drop frames even with significant GPU/CPU/RAM to spare (with high performance plan, not overheating, etc.). Sure, I could just throw money at the problem and upgrade, but I'm yet to find a game I'd want to play bad enough that'd require me to do so. That said, I'd definitely keep VRto3D/WW in mind for desperate times.👌

**Comment by evilkermitreturns** [comment]
@ThreeDeeJay hey read through the thread and thought these might be useful, acer has a GPL notice page at https://images.acer.com/is/content/acer/GPL-Notice_us-enpdf and a software license info doc at https://global-download.acer.com/supportfiles/files/gpl/doc/acer_software_license_information.pdf, the written request path does exist even if i dont think theres any way to find these as a front facing customer, might be worth referencing if you end up pushing for the geo-11/12 source release

**Comment by ThreeDeeJay** [comment]
@evilkermitreturns Yeah, I mentioned it here https://github.com/bo3b/3Dmigoto/issues/354#issuecomment-3534647690
But I'm still not sure if it only applies to customers or if it even applies to geo-11/12 because they've clearly went out of their way to obfuscate the build, configs and even fixes. So I think that even made it really difficult (if even possible) for legit customers to use geo-11/12 in games with custom/no fixes.

But anyhow, I've been told that they supposedly replaced (or plan to?) the offending library (Deviare) ~~so newer builds could be re-licensed under a permissive license like MIT (given the replacement's license allows it) which allows commercial usage without requiring disclosing the source code.~~ Turns out this may not actually be legally sound either since I've been told they can't just re-license 3DMigoto using other code from contributors without their explicit permission, so perhaps the only way they could legally use MIT would be re-basing geo-11 on the last MIT version, before Deviare was added, and probably re-implement all the changes since then (like from over a decade ago) which doesn't seem very likely.

But even if that's true, it begs the question: what about current/older builds? Wouldn't they still be required to release the source code for the builds where they did use Deviare, thus being subject to GPL? 🤔

**Comment by effcol** [comment]
I sent Acer a letter requesting access to their GPL source code in their TrueGame application / Geo-11 / Geo-12, following their [GPL request guidelines on their website.](https://images.acer.com/is/content/acer/GPL-Notice_us-enpdf)

I just received an email, just under two weeks after they received my letter, saying they are working to provide the information I requested and will provide a full update as soon as possible.

I'm feeling pretty positive about this! That feels like about the right amount of time for their legal department to have looked over the request, concluded it's legitimate, and then reached out to say it's progressing while they get the code ready to send.

<img width="1105" height="681" alt="Image" src="https://github.com/user-attachments/assets/dcc9181a-5617-4d4f-9c9e-0e33ce8bf9a5" />

**Comment by effcol** [comment]
Well, it's been about 6 months since my last communication with Acer, and they haven't replied to my follow-up email asking about the status of my GPL request. It seems their compliance department has completely stalled.

With that in mind, I am officially asking the maintainers here to release the open-source code that is under your direct control on binaries you have released.

To @bo3b @DarkStarSword and any maintainers of 3DMigoto, Geo-11, and Geo-12:
Because Geo-11 binaries were actively distributed to the community, there is a clear legal obligation under the GPLv3 to make the corresponding source code public. Since you both maintain access to the repository, could you please open-source the versions of the Geo-11 source code for the binaries you have released? If you have released any Geo-12 binaries as well, please release the source for those too. If you have not released a binary, I believe this does not apply to you (DarkStar, I think you mentioned you hadn't released any binaries).

Apologies for asking so formally. I genuinely don't want to burn any bridges with you. You've been a massive help to the community and to my own projects like wiz3D. If it helps maintain your amicable relationship with Acer, please feel free to wholly blame me and this formal public request for forcing your hand on the Geo-11 release. We've given Acer six months plus of leeway, and it's time to release what we can.

@davegl1234 I am explicitly not asking you to release the Geo-12 source code yourself, as I know you are an Acer contractor/employee and that legal obligation rests solely on Acer as the distributor. However, since you are deeply involved with the Acer SpatialLabs team, could you please help flag this internally and check on the stalled GPL request I submitted with Acer's compliance/product department, evidenced in my posts in the thread above?

The community has formally requested the Geo-11 and Geo-12 source for the binaries Acer shipped, and your help unblocking this internally would be massively appreciated. If it cannot be released, some clarity from Acer as to why would be deeply appreciated.

Thank you all for everything you do for this community.
